src/v/kafka/server/atomic_token_bucket.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "base/seastarx.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/util/shared_token_bucket.hh>
+
+#include <chrono>
+#include <numeric>
+#include <vector>
+
+// atomic_token_bucket tracks the rate of a metric over time using a atomic
+// counters allowing, limiting taking tokens from the bucket it goes over the
+// configured rate.
+//
+// atomic_token_bucket -- configures whether the bucket can grow indefinitely
+// (capped_release::no) or if the bucket is capped to a maximal burst rate set
+// by the limit constructor argument (capped_release::yes)
+template<ss::internal::capped_release capped_release>
+class atomic_token_bucket final {
+public:
+    using clock = ss::lowres_clock;
+
+    /// rate -- the steady state rate we want to allow by the token bucket
+    /// limit -- the maximal burst capacity we allow the bucket to increase to
+    /// if the bucket is capped (capped_release::yes), or the maximal increment
+    /// we allow replenishing the bucket with if the bucket is uncapped
+    /// (capped_release::no)
+    /// threshold -- only replenish the bucket when at least this many tokens
+    /// would be added
+    /// start_full -- sets the last threshold timestamp to well into the past to
+    /// replenish the bucket as much as possible on the next call to replenish()
+    atomic_token_bucket(
+      uint64_t rate, uint64_t limit, uint64_t threshold, bool start_full)
+      : _bucket(rate, limit, threshold, start_full) {}
+
+    /// Take `v` tokens from the bucket
+    /// If there are less than `v` tokens in the bucket, this call will still
+    /// succeed, but subsequent calls to calculate_delay() will return a
+    /// positive delay.
+    void record(uint64_t v) { _bucket.grab(v); }
```

Comment:
I was thinking about folding replenish into one of the methods, but I was thinking about folding it into `calculate_delay` rather than `record/grab`. I saw that `snc_quota_manager` calls `replenish` right after grab, but to me it would make more sense to call replenish right before checking for the delay to ensure the delay is up-to-date. What do you think?